sql-plugin/src/main/scala/com/nvidia/spark/rapids/BatchWithPartitionData.scala

```diff
@@ -0,0 +1,468 @@
+   */
+  def mergeWithPartitionData(): ColumnarBatch = {
+    withResource(inputBatch.getColumnarBatch()) { columnarBatch =>
+      if (partitionSchema.isEmpty) {
+        columnarBatch // No partition columns, return the original batch.
+      } else {
+        val partitionColumns = buildPartitionColumns()
+        val numRows = partitionColumns.head.getRowCount.toInt
+        withResource(new ColumnarBatch(partitionColumns.toArray, numRows)) { partBatch =>
+          assert(columnarBatch.numRows == partBatch.numRows)
+          GpuColumnVector.combineColumns(columnarBatch, partBatch)
+        }
+      }
+    }
+  }
+
+  /**
+   * Builds GPU-based partition columns by mapping partitioned data to a sequence of
+   * GpuColumnVectors.
+   *
+   * @return A sequence of GpuColumnVectors, one for each partitioned column in the schema.
+   */
+  private def buildPartitionColumns(): Seq[GpuColumnVector] = {
+    closeOnExcept(new Array[GpuColumnVector](partitionSchema.length)) { partitionColumns =>
+      for ((field, colIndex) <- partitionSchema.zipWithIndex) {
+        val dataType = field.dataType
+        // Create an array to hold the individual columns for each partition.
+        val singlePartCols = partitionedRowsData.safeMap {
+          case PartitionRowData(valueRow, rowNum) =>
+            val singleValue = valueRow.get(colIndex, dataType)
+            withResource(GpuScalar.from(singleValue, dataType)) { singleScalar =>
+              // Create a column vector from the GPU scalar, associated with the row number.
+              ColumnVector.fromScalar(singleScalar, rowNum)
+            }
+        }
+
+        // Concatenate the individual columns into a single column vector.
+        val concatenatedColumn = withResource(singlePartCols) { _ =>
+          if (singlePartCols.length > 1) {
+            ColumnVector.concatenate(singlePartCols: _*)
+          } else {
+            singlePartCols.head.incRefCount()
+          }
+        }
+        partitionColumns(colIndex) = GpuColumnVector.from(concatenatedColumn, field.dataType)
+      }
+      partitionColumns
+    }
+  }
+
+  override def close(): Unit = {
+    inputBatch.close()
+  }
+}
+
+/**
+ * An iterator for merging and retrieving split ColumnarBatches with partition data using
+ * retry framework.
+ */
+class BatchWithPartitionDataIterator(batchesWithPartitionData: Seq[BatchWithPartitionData])
+  extends GpuColumnarBatchIterator(true) {
+  private val retryIter = withRetry(batchesWithPartitionData.toIterator,
+    BatchWithPartitionDataUtils.splitBatchInHalf) { attempt =>
+    attempt.mergeWithPartitionData()
+  }
+
+  override def hasNext: Boolean = retryIter.hasNext
+
+  override def next(): ColumnarBatch = {
+    retryIter.next()
+  }
+
+  override def doClose(): Unit = {
+    batchesWithPartitionData.foreach(_.close())
+  }
+}
+
+
+object BatchWithPartitionDataUtils {
+  private val CUDF_COLUMN_SIZE_LIMIT: Long = new RapidsConf(SQLConf.get).cudfColumnSizeLimit
```

Comment:
This is problematic from a test perspective.  This is a singleton, so once this class is touched (e.g.: from an earlier test) then it will be computed and never computed again.  That means it's very likely the tests don't actually test anything useful in practice because this will likely be the default value, not what is trying to be overridden via a config in the new tests specific to this fix.

Reply:
Replaced it with a function call.